ethers-core/src/utils/mod.rs

```diff
@@ -421,6 +423,24 @@ pub fn to_checksum(addr: &Address, chain_id: Option<u8>) -> String {
     })
 }
 
+/// Parses an [EIP-1191](https://eips.ethereum.org/EIPS/eip-1191) checksum address.
+///
+/// Returns `Ok(address)` if the checksummed address is valid, `Err()` otherwise.
+/// If `chain_id` is `None`, falls back to [EIP-55](https://eips.ethereum.org/EIPS/eip-55) address checksum method
+pub fn parse_checksummed(addr: &str, chain_id: Option<u8>) -> Result<Address, ConversionError> {
+    let addr = addr.strip_prefix("0x").unwrap_or(addr);
+
+    if let Ok(address) = addr.parse::<Address>() {
+        let checksum_addr = to_checksum(&address, chain_id);
+        if checksum_addr.strip_prefix("0x").unwrap_or(&checksum_addr) == addr {
+            return Ok(address)
+        } else {
+            return Err(ConversionError::InvalidAddressChecksum)
+        }
+    }
+    Err(ConversionError::TextTooLong)
```

Comment:
rather than returning text too long, I think we can do the following

1. add a hex error to the ConversionError enum
```
enum ConversionError { 
    ...,
    /// Hex error, usually from address parsing
    #[error(transparent)]
    FromHexError(#[from]hex::FromHexError)
}
```

2. change this function to use a `?`

```
        let addr: Address = addr.strip_prefix("0x").unwrap_or(addr).parse()?
        if checksum_addr.strip_prefix("0x").unwrap_or(&checksum_addr) == addr {
            return Ok(address)
        } else {
            return Err(ConversionError::InvalidAddressChecksum)
        }


```

Reply:
the `FromHexError` is from `rustc_hex` crate, should I add a it?

Reply:
https://github.com/paritytech/parity-common/blob/master/fixed-hash/src/hash.rs#L568-L593

Reply:
leave it to parity to make our lives more complicated 🤣 🙄

I think you should be able to make it work as follows:
1. reference the fully qualified associated type `<Address as std::str::FromStr>::Err` in the enum variant
2. instead of doing a `#[from]` in the error variant, use `map_err` to convert

```
// variant specifies that it takes "whatever error the `parse()` implementation says is the error"
FromHexError(<Address as std::str::FromStr>::Err)

// map err on the first line of the function
let addr: Address = addr.strip_prefix("0x").unwrap_or(addr).parse().map_err(ConversionError::FromHexError)?;
```

Reply:
sorry this is a pain. if it gets too much let me know and i'll do the tidying

Reply:
@prestwich I've add this `FromHexError`